### PR TITLE
SqlClient add concurrent load tests

### PR DESCRIPTION
**Pull request by Wraith2**
When investigating a concurrency bug I had introduced in SqlEnvChange it was [suggested](https://github.com/dotnet/corefx/pull/36735#issuecomment-481369149) that the DataAccessPerformance benchmarking project that I had been using could be transformed into a test to make sure coverage of continuous concurrent loading was covered in case similar issues are added again. This PR adds those tests.

There are two tests, one each for sync and async. Each tests will warmup for 1 second, execute for 60 seconds and create tasks over 4x the physical core count to ensure that there is contention. The data being read is from the sample Northwind database already used in the manual test project and is similar in shape and types to the Fortunes test of the original. The tests have been added in ProviderAgnostic\ReaderTest as TestConcurrentLoadSync and TestConcurrentLoadAsync because they can be re-used by other providers simply by changing the factory parameter.

/cc @afsanehr, @tarikulsabbir, @Gary-Zh , @david-engel, @saurabh500 

this should unblock https://github.com/dotnet/corefx/pull/36664 and https://github.com/dotnet/corefx/pull/36108

### CODE REVIEW

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs [hunk 1]
```diff
@@ -266,5 +270,371 @@ public static void TestMain()
...
+            private DateTime _stopTime;
+
+            public ConcurrentLoadContext(DbProviderFactory providerFactory, string connectionString, Mode mode, int warmupSeconds, int executionSeconds, int threadCount)
+            {
+                _providerFactory = providerFactory;
+                _connectionString = connectionString;
+                _threadCount = threadCount;
+                _warmupSeconds = warmupSeconds;
+                _executionSeconds = executionSeconds;
+                _query = "SELECT CategoryID,Description FROM Categories";
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]:
```suggestion
                _query = "SELECT CategoryID, Description FROM Categories";
```

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs [hunk 2]
```diff
@@ -266,5 +270,371 @@ public static void TestMain()
...
+                _mode = mode;
+                _start = Start;
+                _stop = Stop;
+                switch (_mode)
+                {
+                    case Mode.Sync:
+                        _work = DoWorkSync;
+                        break;
+                    case Mode.Async:
+                        _work = DoWorkSync;
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: `DoWorkSync` in Async mode ?

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs [hunk 3]
```diff
@@ -266,5 +270,371 @@ public static void TestMain()
...
+            private async Task Stop()
+            {
+                await Task.Delay(TimeSpan.FromSeconds(_warmupSeconds + _executionSeconds));
+                Interlocked.Exchange(ref _running, 0);
+                _stopTime = DateTime.UtcNow;
+            }
+
+            private IEnumerable<Task> CreateTasks()
+            {
+                //yield return Task.Run(
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: Can this code be removed ?

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs [hunk 4]
```diff
@@ -266,5 +270,371 @@ public static void TestMain()
...
+
+                //            lastDisplay = now;
+                //            _totalTransactions += Interlocked.Exchange(ref _counter, 0);
+                //        }
+                //    }
+                //    );
+
+                yield return Task.Run(_start);
+
+                //yield return Task.Run(
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: Commented out code.

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs [hunk 5]
```diff
@@ -266,5 +270,371 @@ public static void TestMain()
...
+                executionSeconds: 60,
+                threadCount: Environment.ProcessorCount * 4
+            );
+            var (transactionPerSecond, average, stdDeviation) = await context.Run();
+            Assert.InRange(transactionPerSecond, 1, int.MaxValue);
+            Assert.True(average > 0);
+            Assert.True(stdDeviation != 0);
+        }
+
+        public class ConcurrentLoadContext
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: This class can be moved to another file altogether.

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs [hunk 6]
```diff
@@ -266,5 +270,371 @@ public static void TestMain()
...
             }
         }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup))]
+        public static async Task TestConcurrentLoadSync()
+        {
+            ConcurrentLoadContext context = new ConcurrentLoadContext(
+                providerFactory: SqlClientFactory.Instance,
+                connectionString: DataTestUtility.TcpConnStr,
+                mode: ConcurrentLoadContext.Mode.Async,
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]:
```suggestion
                mode: ConcurrentLoadContext.Mode.Sync,
```
Passing async mode in sync test ?

Comment (Wraith2) [reply src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: copy paste error, fixed and retested.

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs (hunk 5 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: done.

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs (hunk 4 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: all removed

File: src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs (hunk 1 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/tests/ManualTests/ProviderAgnostic/ReaderTest/ReaderTest.cs]: done.

**Comment by saurabh500** [comment]
@AfsanehR  will put this through the CI and merge.

**Comment by AfsanehR** [comment]
Successfully passed the tests.

**Pull request merged**